### PR TITLE
feat: control the proxy_upstream in lua side

### PR DESCRIPTION
**Pull request by oowl**
Set upstream next enablement of current request to the given string of table
argument . Global setting set by [`proxy_next_upstream`](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_next_upstream) will be overwritten.

[AG-186](https://konghq.atlassian.net/browse/AG-186)

[AG-186]: https://konghq.atlassian.net/browse/AG-186?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/ngx_http_lua_kong_module.c [hunk 1]
```diff
@@ -158,4 +158,75 @@ ngx_http_lua_kong_merge_loc_conf(ngx_conf_t *cf, void *parent, void *child)
...
+#define NGX_HTTP_UPSTREAM_FT_STATUS          (NGX_HTTP_UPSTREAM_FT_HTTP_500  \
+                                             |NGX_HTTP_UPSTREAM_FT_HTTP_502  \
+                                             |NGX_HTTP_UPSTREAM_FT_HTTP_503  \
+                                             |NGX_HTTP_UPSTREAM_FT_HTTP_504  \
+                                             |NGX_HTTP_UPSTREAM_FT_HTTP_403  \
+                                             |NGX_HTTP_UPSTREAM_FT_HTTP_404  \
+                                             |NGX_HTTP_UPSTREAM_FT_HTTP_429)
+
+static ngx_conf_bitmask_t  ngx_http_proxy_next_upstream_masks[] = {
+    { ngx_string("error"), NGX_HTTP_UPSTREAM_FT_ERROR },
```
Comment (fffonion) [inline src/ngx_http_lua_kong_module.c:187]: seems this is not used

File: lualib/resty/kong/upstream.lua [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+function _M.set_next_upstream(...)
+	local nargs = select("#", ...)
+	if nargs == 0 then
+		return "no argument"
+	end
+
+	local r = get_request()
+	if not r then
+		return "no request found"
+	end
```
Comment (tzssangglass) [inline lualib/resty/kong/upstream.lua:57]: indentation

File: lualib/resty/kong/upstream.lua [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	http_404 = 0x00000200,
+	http_429 = 0x00000400,
+	off = 0x00001000,
+	non_idempotent = 0x00004000,
+}
+
+function _M.set_next_upstream(...)
+	local nargs = select("#", ...)
+	if nargs == 0 then
+		return "no argument"
```
Comment (tzssangglass) [inline lualib/resty/kong/upstream.lua:51]: indentation

File: lualib/resty/kong/upstream.lua [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		if type(v) ~= "string" then
+			return "argument #" .. i .. " is not a string"
+		end
+
+		local next_upstream_value = next_upstream_table[v]
+		if not next_upstream_value then
+			return "argument #" .. i .. " is not a valid argument"
+		end
+
+		next_upstream = bit.bor(next_upstream, next_upstream_value)
```
Comment (tzssangglass) [inline lualib/resty/kong/upstream.lua:72]: bad indentation in this file

File: lualib/resty/kong/upstream.lua [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	http_504 = 0x00000080,
+	http_403 = 0x00000100,
+	http_404 = 0x00000200,
+	http_429 = 0x00000400,
+	off = 0x00001000,
+	non_idempotent = 0x00004000,
+}
+
+function _M.set_next_upstream(...)
+	local nargs = select("#", ...)
```
Comment (tzssangglass) [inline lualib/resty/kong/upstream.lua:49]: local select

File: src/ngx_http_lua_kong_module.c [hunk 6]
```diff
@@ -158,4 +158,33 @@ ngx_http_lua_kong_merge_loc_conf(ngx_conf_t *cf, void *parent, void *child)
     return NGX_CONF_OK;
 }
 
+ngx_flag_t
+ngx_http_lua_kong_get_next_upstream_mask(ngx_http_request_t *r,
+    ngx_flag_t upstream_next)
```
Comment (tzssangglass) [inline src/ngx_http_lua_kong_module.c:163]: It seems this function is not used?

Comment (oowl) [reply src/ngx_http_lua_kong_module.c:163]: https://github.com/Kong/kong/pull/14029/files#diff-87978139f4c7701a35f1c2eaf5b8f71ccc16f9549571ab6c97dc127ba6be31bbR10 in here

File: lualib/resty/kong/upstream.lua (hunk 3 above)
Comment (oowl) [reply lualib/resty/kong/upstream.lua:51]: indentation is 4 spaces, what's the problem here?

File: lualib/resty/kong/upstream.lua (hunk 5 above)
Comment (oowl) [reply lualib/resty/kong/upstream.lua:50]: done

**Review by tzssangglass — APPROVED** [review]

**Review by fffonion — APPROVED** [review]

**Pull request merged**